File: 67. Add Binary/Solution.java
class Solution {
    public String addBinary(String a, String b) {
        StringBuilder str = new StringBuilder();
        int m = a.length(), n = b.length(), carry = 0, total;
        for (int i = 1; i <= Math.min(m, n); i++) {
            total = a.charAt(m - i) - '0' + b.charAt(n - i) - '0' + carry;
            str.append(total % 2);
            carry = total > 1 ? 1 : 0;
        }
        for (int i = Math.min(m, n) + 1; i <= m; i++) {
            total = a.charAt(m - i) - '0' + carry;
            str.append(total % 2);
            carry = total > 1 ? 1 : 0;
        }
        for (int i = Math.min(m, n) + 1; i <= n; i++) {
            total = b.charAt(n - i) - '0' + carry;
            str.append(total % 2);
            carry = total > 1 ? 1 : 0;
        }
        if (carry > 0) str.append(1);
        return str.reverse().toString();
    }
}

File: 67. Add Binary/Solution.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        res = []
        m, n, carry, total = len(a), len(b), 0, 0
        for i in range(1, min(m, n) + 1):
            total = ord(a[m - i]) + ord(b[n - i]) - 2 * ord('0') + carry
            res.append(str(total % 2))
            carry = 1 if total > 1 else 0
        for i in range(min(m, n) + 1, m + 1):
            total = ord(a[m - i]) - ord('0') + carry
            res.append(str(total % 2))
            carry = 1 if total > 1 else 0
        for i in range(min(m, n) + 1, n + 1):
            total = ord(b[n - i]) - ord('0') + carry
            res.append(str(total % 2))
            carry = 1 if total > 1 else 0
        if carry > 0: res.append(str(1))
        return ''.join(reversed(res))